File: scripts/smartcity-gps-sync.py
#!/usr/bin/env python3
"""
SmartCity GPS Sync Script
Ovaj skript se pokreće na legacy serveru kao CRON job
i šalje GPS podatke na SmartCity API

CRON setup: */5 * * * * /usr/bin/python3 /path/to/smartcity-gps-sync.py

Instalacija dependency-ja:
pip install mysql-connector-python requests
"""

import json
import logging
import sys
from datetime import datetime, timedelta
from typing import List, Dict, Any

import mysql.connector
import requests
from mysql.connector import Error

# Konfiguracija
CONFIG = {
    'api_url': 'https://api.smart-city.rs/gps-ingest/batch',  # Promeni na produkcijski URL
    'api_key': 'smartcity_legacy_gps_key_2024',
    'db_host': '79.101.48.11',
    'db_port': 3306,
    'db_name': 'pib100065430gps',
    'db_user': 'YOUR_DB_USER',  # Postavi kredencijale
    'db_pass': 'YOUR_DB_PASS',
    'batch_size': 1000,  # Broj GPS tačaka po batch-u
    'time_window_minutes': 5,  # Poslednih 5 minuta
    'log_file': '/var/log/smartcity-gps-sync.log',
    'debug': True,
    'ssl_verify': False  # Za development, u produkciji postaviti na True
}

# Setup logging
logging.basicConfig(
    level=logging.DEBUG if CONFIG['debug'] else logging.INFO,
    format='[%(asctime)s] [%(levelname)s] %(message)s',
    handlers=[
        logging.FileHandler(CONFIG['log_file']),
        logging.StreamHandler(sys.stdout)
    ]
)
logger = logging.getLogger(__name__)


def fetch_gps_data(connection) -> List[Dict[str, Any]]:
    """Dohvati GPS podatke iz legacy baze"""
    try:
        cursor = connection.cursor(dictionary=True)
        
        # Kalkuliši vremenski okvir
        time_threshold = datetime.now() - timedelta(minutes=CONFIG['time_window_minutes'])
        
        query = """
            SELECT 
                garageNo,
                lat,
                lng,
                speed,
                course,
                alt,
                state,
                inroute AS inRoute,
                line_number AS lineNumber,
                direction,
                captured,
                edited,
                (IFNULL(people_counter_1_in, 0) + IFNULL(people_counter_2_in, 0) + 
                 IFNULL(people_counter_3_in, 0) + IFNULL(people_counter_4_in, 0)) AS peopleIn,
                (IFNULL(people_counter_1_out, 0) + IFNULL(people_counter_2_out, 0) + 
                 IFNULL(people_counter_3_out, 0) + IFNULL(people_counter_4_out, 0)) AS peopleOut,
                battery_status AS batteryStatus
            FROM current
            WHERE captured > %s
            AND garageNo IS NOT NULL
            AND lat IS NOT NULL
            AND lng IS NOT NULL
            ORDER BY captured DESC
            LIMIT %s
        """
        
        cursor.execute(query, (time_threshold, CONFIG['batch_size']))
        results = cursor.fetchall()
        cursor.close()
        
        logger.info(f"Fetched {len(results)} GPS points from database")
        return results
        
    except Error as e:
        logger.error(f"Database error: {e}")
        raise


def prepare_payload(gps_data: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Pripremi podatke za slanje na API"""
    
    def format_point(row):
        """Formatiraj pojedinačnu GPS tačku"""
        return {
            'garageNo': row['garageNo'],
            'lat': float(row['lat']),
            'lng': float(row['lng']),
            'speed': int(row['speed'] or 0),
            'course': int(row['course'] or 0),
            'alt': int(row['alt'] or 0),
            'state': int(row['state'] or 0),
            'inRoute': int(row['inRoute'] or 0),
            'lineNumber': row['lineNumber'],
            'direction': int(row['direction']) if row['direction'] else None,
            'peopleIn': int(row['peopleIn'] or 0),
            'peopleOut': int(row['peopleOut'] or 0),
            'batteryStatus': int(row['batteryStatus']) if row['batteryStatus'] else None,
            'captured': row['captured'].isoformat() if isinstance(row['captured'], datetime) else str(row['captured']),
            'edited': row['edited'].isoformat() if isinstance(row['edited'], datetime) else str(row['edited'])
        }
    
    return {
        'data': [format_point(row) for row in gps_data],
        'source': 'legacy_cron',
        'timestamp': datetime.now().isoformat()
    }


def send_to_api(payload: Dict[str, Any]) -> Dict[str, Any]:
    """Pošalji podatke na SmartCity API"""
    headers = {
        'Content-Type': 'application/json',
        'X-API-Key': CONFIG['api_key']
    }
    
    try:
        response = requests.post(
            CONFIG['api_url'],
            json=payload,
            headers=headers,
            timeout=30,
            verify=CONFIG['ssl_verify']
        )
        
        response.raise_for_status()
        return response.json()
        
    except requests.RequestException as e:
        logger.error(f"API request failed: {e}")
        if hasattr(e, 'response') and e.response is not None:
            logger.error(f"Response: {e.response.text}")
        raise


def main():
    """Glavni program"""
    logger.info("=== SmartCity GPS Sync Started ===")
    
    connection = None
    try:
        # Konektuj se na bazu
        logger.info("Connecting to legacy database...")
        connection = mysql.connector.connect(
            host=CONFIG['db_host'],
            port=CONFIG['db_port'],
            database=CONFIG['db_name'],
            user=CONFIG['db_user'],
            password=CONFIG['db_pass'],
            connection_timeout=10
        )
        
        if not connection.is_connected():
            raise Exception("Failed to connect to database")
        
        logger.info("Connected to legacy database")
        
        # Dohvati GPS podatke
        gps_data = fetch_gps_data(connection)
        
        if not gps_data:
            logger.info("No new GPS data to sync")
            return 0
        
        # Pripremi i pošalji podatke
        payload = prepare_payload(gps_data)
        logger.info(f"Sending {len(gps_data)} GPS points to API...")
        
        result = send_to_api(payload)
        
        if result.get('success'):
            logger.info(f"Successfully synced {result.get('processed', 0)} GPS points")
            if result.get('failed', 0) > 0:
                logger.warning(f"Failed to sync {result['failed']} GPS points")
        else:
            logger.error(f"API returned error: {result}")
            return 1
            
    except Exception as e:
        logger.error(f"Fatal error: {e}")
        return 1
        
    finally:
        if connection and connection.is_connected():
            connection.close()
            logger.info("Database connection closed")
    
    logger.info("=== SmartCity GPS Sync Completed ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/smartcity-gps-sync.php
<?php
/**
 * SmartCity GPS Sync Script
 * Ovaj skript se pokreće na legacy serveru kao CRON job
 * i šalje GPS podatke na SmartCity API
 * 
 * CRON setup: */5 * * * * /usr/bin/php /path/to/smartcity-gps-sync.php
 */

// Konfiguracija
$config = [
    'api_url' => 'https://api.smart-city.rs/gps-ingest/batch', // Promeni na produkcijski URL
    'api_key' => 'smartcity_legacy_gps_key_2024',
    'db_host' => '79.101.48.11',
    'db_port' => 3306,
    'db_name' => 'pib100065430gps',
    'db_user' => 'YOUR_DB_USER', // Postavi kredencijale
    'db_pass' => 'YOUR_DB_PASS',
    'batch_size' => 1000, // Broj GPS tačaka po batch-u
    'time_window' => 300, // Poslednih 5 minuta (300 sekundi)
    'log_file' => '/var/log/smartcity-gps-sync.log',
    'debug' => true
];

// Logging funkcija
function logMessage($message, $level = 'INFO') {
    global $config;
    $timestamp = date('Y-m-d H:i:s');
    $logLine = "[$timestamp] [$level] $message\n";
    
    if ($config['debug']) {
        echo $logLine;
    }
    
    if ($config['log_file']) {
        file_put_contents($config['log_file'], $logLine, FILE_APPEND);
    }
}

// Početak skripte
logMessage("=== SmartCity GPS Sync Started ===");

try {
    // Konektuj se na legacy bazu
    $dsn = "mysql:host={$config['db_host']};port={$config['db_port']};dbname={$config['db_name']};charset=utf8mb4";
    $pdo = new PDO($dsn, $config['db_user'], $config['db_pass']);
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    
    logMessage("Connected to legacy database");
    
    // Dohvati GPS podatke iz current tabele (poslednih 5 minuta)
    $sql = "
        SELECT 
            garageNo,
            lat,
            lng,
            speed,
            course,
            alt,
            state,
            inroute AS inRoute,
            line_number AS lineNumber,
            direction,
            captured,
            edited,
            people_counter_1_in + people_counter_2_in + people_counter_3_in + people_counter_4_in AS peopleIn,
            people_counter_1_out + people_counter_2_out + people_counter_3_out + people_counter_4_out AS peopleOut,
            battery_status AS batteryStatus
        FROM current
        WHERE captured > DATE_SUB(NOW(), INTERVAL :time_window SECOND)
        AND garageNo IS NOT NULL
        AND lat IS NOT NULL
        AND lng IS NOT NULL
        ORDER BY captured DESC
        LIMIT :batch_size
    ";
    
    $stmt = $pdo->prepare($sql);
    $stmt->execute([
        'time_window' => $config['time_window'],
        'batch_size' => $config['batch_size']
    ]);
    
    $gpsData = $stmt->fetchAll(PDO::FETCH_ASSOC);
    $count = count($gpsData);
    
    logMessage("Found $count GPS points to sync");
    
    if ($count == 0) {
        logMessage("No new GPS data to sync");
        exit(0);
    }
    
    // Pripremi podatke za slanje
    $payload = [
        'data' => array_map(function($row) {
            return [
                'garageNo' => $row['garageNo'],
                'lat' => (float)$row['lat'],
                'lng' => (float)$row['lng'],
                'speed' => (int)$row['speed'],
                'course' => (int)$row['course'],
                'alt' => (int)($row['alt'] ?? 0),
                'state' => (int)($row['state'] ?? 0),
                'inRoute' => (int)($row['inRoute'] ?? 0),
                'lineNumber' => $row['lineNumber'],
                'direction' => $row['direction'] ? (int)$row['direction'] : null,
                'peopleIn' => (int)($row['peopleIn'] ?? 0),
                'peopleOut' => (int)($row['peopleOut'] ?? 0),
                'batteryStatus' => $row['batteryStatus'] ? (int)$row['batteryStatus'] : null,
                'captured' => date('c', strtotime($row['captured'])),
                'edited' => date('c', strtotime($row['edited']))
            ];
        }, $gpsData),
        'source' => 'legacy_cron',
        'timestamp' => date('c')
    ];
    
    // Pošalji podatke na API
    $ch = curl_init($config['api_url']);
    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode($payload),
        CURLOPT_HTTPHEADER => [
            'Content-Type: application/json',
            'X-API-Key: ' . $config['api_key']
        ],
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT => 30,
        CURLOPT_SSL_VERIFYPEER => false // Za development, u produkciji postaviti na true
    ]);
    
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);
    
    if ($error) {
        throw new Exception("CURL error: $error");
    }
    
    if ($httpCode !== 200) {
        throw new Exception("API returned HTTP $httpCode: $response");
    }
    
    $result = json_decode($response, true);
    
    if ($result && $result['success']) {
        logMessage("Successfully synced {$result['processed']} GPS points");
        if ($result['failed'] > 0) {
            logMessage("Failed to sync {$result['failed']} GPS points", 'WARNING');
        }
    } else {
        throw new Exception("API error: $response");
    }
    
} catch (Exception $e) {
    logMessage("ERROR: " . $e->getMessage(), 'ERROR');
    exit(1);
}

logMessage("=== SmartCity GPS Sync Completed ===");
exit(0);